admin/test/test_release.py

```diff
@@ -661,3 +667,637 @@ def test_publish_non_release_fails(self):
+        }
+
+        aws = FakeAWS(
+            routing_rules={},
+            s3_buckets={
+                self.target_bucket: existing_s3_keys.copy(),
+            },
+        )
+
+        self.update_repo(
+            aws=aws,
+            yum=FakeYum(),
+            rpm_directory=self.rpm_directory,
+            target_bucket=self.target_bucket,
+            target_key=self.target_key,
+            source_repo=self.create_fake_repository(files={}),
+            packages=self.packages,
+        )
+
+        self.assertNotEqual(
+            existing_s3_keys[index_path],
+            aws.s3_buckets[self.target_bucket][index_path])
+
+    def test_existing_metadata_files_not_uploaded(self):
+        """
+        Calling :func:`update_repo` does not update repository metadata files
+        which are not the index.
+        """
+        index_path = os.path.join(self.target_key, 'repodata', 'repomd.xml')
+        existing_metadata_file = os.path.join(self.target_key, 'repodata',
+                                              'filelists.xml.gz')
+
+        existing_s3_keys = {
+            index_path: 'old_metadata_index',
+            existing_metadata_file: 'old_metadata_content',
+        }
+
+        aws = FakeAWS(
+            routing_rules={},
+            s3_buckets={
+                self.target_bucket: existing_s3_keys.copy(),
+            },
+        )
+
+        self.update_repo(
+            aws=aws,
+            yum=FakeYum(),
+            rpm_directory=self.rpm_directory,
+            target_bucket=self.target_bucket,
+            target_key=self.target_key,
+            source_repo=self.create_fake_repository(files={}),
+            packages=self.packages,
+        )
+
+        self.assertEqual(
+            aws.s3_buckets[self.target_bucket][existing_metadata_file],
+            existing_s3_keys[existing_metadata_file])
+
+    def test_new_metadata_files_uploaded(self):
+        """
+        Calling :func:`update_repo` uploads new repository metadata files to
+        S3.
+        """
+        index_path = os.path.join(self.target_key, 'repodata', 'repomd.xml')
+        existing_s3_keys = {
+            index_path: 'old_metadata_index',
+        }
+
+        aws = FakeAWS(
+            routing_rules={},
+            s3_buckets={
+                self.target_bucket: existing_s3_keys.copy(),
+            },
+        )
+
+        self.update_repo(
+            aws=aws,
+            yum=FakeYum(),
+            rpm_directory=self.rpm_directory,
+            target_bucket=self.target_bucket,
+            target_key=self.target_key,
+            source_repo=self.create_fake_repository(files={}),
+            packages=self.packages,
+        )
+
+        repodata_files = [
+            key for key in aws.s3_buckets[self.target_bucket] if
+            key.startswith(os.path.join(self.target_key, 'repodata'))]
+
+        # What matters is that there is more than just the index.
+        self.assertGreater(len(repodata_files), 1)
```

Comment:
We should check at least one of the metadata files (although, it would probably be enough to have one fake one (but with a hash in the filename).